distributed/worker.py

```diff
@@ -610,23 +612,26 @@ def _register_with_scheduler(self):
             raise ValueError("Unexpected response from register: %r" %
                              (response,))
         else:
-            # Retrieve eventual init functions and run them
-            for function_bytes in response['worker-setups']:
-                setup_function = pickle.loads(function_bytes)
-                if has_arg(setup_function, 'dask_worker'):
-                    result = setup_function(dask_worker=self)
-                else:
-                    result = setup_function()
-                logger.info('Init function %s ran: output=%s' % (setup_function, result))
-
             logger.info('        Registered to: %26s', self.scheduler.address)
             logger.info('-' * 49)
 
+        # Retrieve eventual init functions (only worker-setups for now)
+        for name, function_bytes in response['worker-setups'].items():
+            self.worker_setups[name] = pickle.loads(function_bytes)
+
         self.batched_stream = BatchedSend(interval='2ms', loop=self.loop)
         self.batched_stream.start(comm)
         self.periodic_callbacks['heartbeat'].start()
         self.loop.add_callback(self.handle_scheduler, comm)
 
+        # run eventual init functions (only worker-setups for now)
```

Comment:
I moved this to run the worker init functions after heartbeat; also I am using the locak worker_setups directionary rather than the unpickled function directly.